WEBGL_get_buffer_sub_data_async extension

kainino0x opened the pull request:
I took @kenrussell's spec text from #1937 but significantly rewrote it text to exactly match the new text for getBufferSubData where appropriate. I also changed the IDL to match getBufferSubData.

This entry point solves the problem of synchronous, blocking GPU-to-CPU memory readback on the main thread.

Some performance results in real-world usage (screen picking):
https://github.com/kainino0x/getBufferSubDataAsync-Demo
- [x] Copy rewording from #2078 once it's settled
- [x] Don't test ordering: https://github.com/KhronosGroup/WebGL/pull/2079#discussion_r82401676
- [x] Add test for the transform feedback error case to the test added in #2097 (it should be added to `verifyGetBufferSubData`)
- [x] Move to an extension
- [x] Remove unnecessary manual tests in `sdk/tests/extra/`.

This PR fixes #1938.

kenrussell commented on sdk/tests/conformance2/buffers/getBufferSubDataAsync-stress.html:
```diff
@@ -0,0 +1,98 @@
+for (var i = 0; i < TEST_COUNT; i++) {
+  (function() {
+    var counter = i;
+    var color = [counter & 0xff, (counter << 8) & 0xff, (counter << 16) & 0xff, 255];
```
Did you mean to right shift instead of left shift these? They'll always shift in zero bits as written.

kenrussell replied:
Still looking for an update in this area.

kenrussell commented on sdk/tests/conformance2/buffers/getBufferSubDataAsync.html:
```diff
@@ -0,0 +1,249 @@
+  debug("");
+  finishTest();
+}).catch(function(e) {
+  testFailed(e.toString());
```
Should this also call finishTest so the test harness (webgl-conformance-test.html) doesn't report this as a timeout if an exception is thrown?

Kangz commented on sdk/tests/conformance2/buffers/getBufferSubDataAsync-stress.html:
```diff
@@ -0,0 +1,98 @@
+            buf + " but expected " + color);
+      }
+      if (counter <= last_counter) {
+        testFailed("getBufferSubDataAsync calls #" + last_counter + " and #" + counter + " out of order");
```
The getBufferSubDataAsync promises being called in order doesn't appear in the spec (or I didn't find it) and seems to be an implementation detail. How about having a vector of bools and in Promise.all(promises).then() check that it is all true?

kenrussell replied:
I'm not sure how the guarantees about what data is seen in the ArrayBufferView would be maintained if the promises resolved out of order. The same ArrayBufferView might be used as the target to getBufferSubData multiple times, and the promises' "then" callbacks would assume the data would be observed sequentially.

Should we say something in the spec about sequential calls A and B to getBufferSubDataAsync, and that the promise for call A must be resolved before the promise to call B, assuming both are not rejected early?

It'd be fine by me to change the test. Is there a simple way to express success? I would assume that Promise.all's resolve callback would only be called upon success of all of the individual promises and their resolve callbacks -- so throwing an exception out of one of them would cause Promise.all to fail. Could you investigate?

kainino0x replied:
Since the ArrayBufferView isn't written until promise resolution, I don't see a problem there. We don't make any guarantees about the contents of the ArrayBufferView after the promise callback is executed; only during its execution.

Yes, failing one would cause Promise.all to fail. But there can be `testFailed` messages inside the individual callbacks as well so I think that would be okay.

Kangz replied:
Like Kai, I don't think we need to add ordering to the spec, the fact that all the promises are resolved on the borwsing context event loop gives the guarantee that inside the promise.then, the ArrayBufferView content is correct.

kenrussell replied:
OK. Given your analyses I think it's fine to leave the spec as is.

Kangz commented on sdk/tests/conformance2/buffers/getBufferSubDataAsync.html:
```diff
@@ -0,0 +1,249 @@
+});
+
+allPromises = allPromises.then(function() {
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, new ArrayBuffer(4))";
```
Why should these be rejected? Is it because no buffer is bound to gl.ARRAY_BUFFER (I think some might be)?

kainino0x replied:
Those are type checks, which I didn't make clear at all. I added debug prints to explain.

```
PASS correctly rejected with TypeError: Failed to execute 'getBufferSubDataAsync' on 'WebGL2RenderingContext': parameter 3 is not of type 'ArrayBufferView'.: promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, new ArrayBuffer(4))
```

```
PASS correctly rejected with TypeError: Failed to execute 'getBufferSubDataAsync' on 'WebGL2RenderingContext': parameter 3 is not of type 'ArrayBufferView'.: promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, null)
```

Kangz commented on sdk/tests/conformance2/buffers/getBufferSubDataAsync.html:
```diff
@@ -0,0 +1,249 @@
+  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, code);
+  return expectRejected(promise, code);
+});
+
```
A couple more tests that might be useful:
 - Check that getBufferSubDataAsync on an uninitialized (gl.bufferData with null) buffer returns zeros.
 - Check that getBufferSubDataAsync immediately followed by a buffer resize to be too small doesn't crash.

kainino0x commented on specs/latest/2.0/index.html:
```diff
@@ -1216,6 +1219,77 @@ <h2 class="no-toc">Table of contents</h2>
+        <code>dstData</code>. If any of the above errors occur, the Promise is already rejected upon
+        return. If the context is lost before the Promise is fulfilled, or if <code>dstData</code>
+        is neutered at the time the Promise is fulfilled, then the Promise will be rejected later.
+        In this case, no WebGL error is generated. If the promise rejects, no data is written to
```
@kenrussell Just want to make sure that this is the semantic we want. Right now no WebGL error will ever be generated upon resolution of the promise.

kenrussell replied:
Yes, that sounds like the best semantic to me.

kainino0x commented:
I merged master into this branch, updated the spec language to match getBufferSubData, etc. I think I've addressed all of the outstanding comments and issues.

Kangz commented:
Still LGTM

kainino0x commented:
Added a test for the transform feedback error case (finally).

Also added in getBufferSubData tests in #2097

kainino0x commented:
> Added a test for the transform feedback error case (finally).

Removed, will add it back to this PR once that one has landed and this PR's fate has stabilized.

kenrussell commented:
Excellent work. For a little more background for other reviewers: the new form of the spec text came from a suggestion from Elliott Sprehn from the Blink team, pointing us at other web specs that return promises.

LGTM.

kainino0x commented:
I've finally moved this entire spec into an extension. The extension is implemented and tested on my local Chromium build. Aside from the new specification text, everything is as it was before. Ready for review.

zhenyao commented on sdk/tests/conformance2/extensions/webgl-get-buffer-sub-data-async-stress.html:
```diff
@@ -0,0 +1,111 @@
+  var promises = [];
+  debug("kicking off " + TEST_COUNT + " getBufferSubDataAsync calls");
+  for (var i = 0; i < TEST_COUNT; i++) {
+    (function() {
```
Curious: why make the loop body a function? why regular code will not do?

kainino0x replied:
This is a javascript programming pattern that guards against a weirdness of closures, which is that they always capture variables, not values. So if you capture `i`, then all of the closures (promise callbacks) will see `i == TEST_COUNT` instead of the loop value.

After writing this, though, I learned about `let`, which actually will solve this. I'm going to go ahead and clean this up using `let`.

zhenyao commented on sdk/tests/conformance2/extensions/webgl-get-buffer-sub-data-async.html:
```diff
@@ -0,0 +1,330 @@
+  });
+
+  for (var i = 0; i < types.length; i++) {
+    (function() {
```
Same question here.

zhenyao commented on sdk/tests/extra/getBufferSubDataAsync-dom-suspension-manual.html:
```diff
@@ -0,0 +1,125 @@
+  debug("kicking off " + TEST_COUNT + " getBufferSubDataAsync calls");
+  for (var i = 0; i < TEST_COUNT; i++) {
+    insertTimeoutAssertion();
+    (function() {
```
Same question here?

kainino0x replied:
I actually forgot about these tests and haven't updated them. I need to do this before I merge.

zhenyao replied:
What about leaving these manual tests out of this pull request?

kainino0x replied:
Sure, good call. Done.

zhenyao commented:
lgtm

kenrussell commented:
Very nice @kainino0x and thanks @zhenyao for carrying forward the code review. LGTM after the fact.
